pkg/util/voljournal.go

```diff
@@ -0,0 +1,394 @@
+			" request name (%s) volume UUID (%s) volume omap name (%s)",
+			reqName, objUUID, savedReqName)
+	}
+
+	if snapSource {
+		// check if source UUID key points back to the parent volume passed in
+		if savedReqParentName != parentName {
+			// NOTE: This can happen if there is a snapname conflict, and we already have a snapshot
+			// with the same name pointing to a different UUID as the source
+			err = fmt.Errorf("snapname points to different volume, request name (%s)"+
+				" source name (%s) saved source name (%s)",
+				reqName, parentName, savedReqParentName)
+			return "", ErrSnapNameConflict{reqName, err}
+		}
+	}
+
+	return objUUID, nil
+}
+
+/*
+UndoReservation undoes a reservation, in the reverse order of ReserveName
+- The UUID directory is cleaned up before the VolName key in the csiDirectory is cleaned up
+
+NOTE: Ensure that the Ceph volume (image or FS subvolume) backing the reservation is cleaned up
+prior to cleaning up the reservation
+
+NOTE: As the function manipulates omaps, it should be called with a lock against the request name
+held, to prevent parallel operations from modifying the state of the omaps for this request name.
+*/
+func (cj *CSIJournal) UndoReservation(monitors, id, key, pool, volName, reqName string) error {
+	// delete volume UUID omap (first, inverse of create order)
+	// TODO: Check cases where volName can be empty, and we need to just cleanup the reqName
+	imageUUID := strings.TrimPrefix(volName, cj.namingPrefix)
+	err := RemoveObject(monitors, id, key, pool, cj.cephUUIDDirectoryPrefix+imageUUID)
+	if err != nil {
+		if _, ok := err.(ErrObjectNotFound); !ok {
+			klog.Errorf("failed removing oMap %s (%s)", cj.cephUUIDDirectoryPrefix+imageUUID, err)
+			return err
+		}
+	}
+
+	// delete the request name key (last, inverse of create order)
+	err = RemoveOMapKey(monitors, id, key, pool, cj.csiDirectory,
+		cj.csiNameKeyPrefix+reqName)
+	if err != nil {
+		klog.Errorf("failed removing oMap key %s (%s)", cj.csiNameKeyPrefix+reqName, err)
+		return err
+	}
+
+	return err
+}
+
+// reserveOMapName creates an omap with passed in oMapNamePrefix and a generated <uuid>.
+// It ensures generated omap name does not already exist and if conflicts are detected, a set
+// number of retires with newer uuids are attempted before returning an error
+func reserveOMapName(monitors, id, key, pool, oMapNamePrefix string) (string, error) {
+	var iterUUID string
+
+	maxAttempts := 5
+	attempt := 1
+	for attempt <= maxAttempts {
+		// generate a uuid for the image name
+		iterUUID = uuid.NewUUID().String()
+
+		err := CreateObject(monitors, id, key, pool, oMapNamePrefix+iterUUID)
+		if err != nil {
+			if _, ok := err.(ErrObjectExists); ok {
+				attempt++
+				// try again with a different uuid, for maxAttempts tries
+				klog.V(4).Infof("uuid (%s) conflict detected, retrying (attempt %d of %d)",
+					iterUUID, attempt, maxAttempts)
+				continue
+			}
+
+			return "", err
+		}
+
+		break
+	}
+
+	if attempt > maxAttempts {
```

Comment:
If we are here, its sure that we exceeded the attemtps, Isnt it ? if then why we need this condition ?

Reply:
The success case was returned below this check, hence this was needed. I now moved the "success return" to the older "break on success" location in the for loop, hence removed this check.